src/Juvix/Core/IR/Evaluator.hs

```diff
@@ -6,177 +6,389 @@
-    IR.Bound' ii _ ->
-      fromMaybe (throw @"typecheckError" (IR.UnboundIndex ii)) $
-        pure |<< (d ^? ix (fromIntegral ii))
-
-evalElim ::
-  forall ext primTy primVal m.
-  ( HasThrow "typecheckError" (IR.TypecheckError primTy primVal m) m,
-    Show primTy,
-    Show primVal,
-    IR.TermX ext primTy primVal ~ Void,
-    IR.ElimX ext primTy primVal ~ Void
-  ) =>
-  EvalElim primTy primVal ext m
-evalElim = evalElimWith absurd absurd
-
--- value application function
-vapp ::
-  forall primTy primVal m.
-  ( HasThrow "typecheckError" (IR.TypecheckError primTy primVal m) m,
-    Show primTy,
-    Show primVal
-  ) =>
-  Parameterisation primTy primVal ->
-  IR.Value primTy primVal m ->
-  IR.Value primTy primVal m ->
-  m (IR.Value primTy primVal m)
-vapp _ (IR.VLam f) v = f v
-vapp _ (IR.VNeutral n) v = pure (IR.VNeutral (IR.NApp n v))
-vapp param (IR.VPrim x) (IR.VPrim y) =
-  case Juvix.Core.Types.apply param x y of
-    Just v -> pure (IR.VPrim v)
-    Nothing -> throw @"typecheckError" (IR.CannotApply (IR.VPrim x) (IR.VPrim y))
-vapp _ f x = throw @"typecheckError" (IR.CannotApply f x)
-
-class CanSubst ext primTy primVal t where
-  subst :: Natural -> IR.Elim' ext primTy primVal -> t -> t
-
-instance CanSubst ext primTy primVal () where subst _ _ = identity
-
-instance CanSubst ext primTy primVal Void where subst _ _ = absurd
-
-instance
-  ( IR.TermAll (CanSubst ext primTy primVal) ext primTy primVal,
-    IR.ElimAll (CanSubst ext primTy primVal) ext primTy primVal
-  ) =>
-  CanSubst ext primTy primVal (IR.Term' ext primTy primVal)
-  where
-  subst ii r (IR.Star' i a) = IR.Star' i (subst ii r a)
-  subst ii r (IR.PrimTy' p a) = IR.PrimTy' p (subst ii r a)
-  subst ii r (IR.Pi' pi ty ty' a) =
-    IR.Pi' pi (subst ii r ty) (subst (ii + 1) r ty') (subst ii r a)
-  subst ii r (IR.Lam' f a) = IR.Lam' (subst (ii + 1) r f) (subst ii r a)
-  subst ii r (IR.Elim' e a) = IR.Elim' (subst ii r e) (subst ii r a)
-  subst ii r (IR.TermX a) = IR.TermX (subst ii r a)
-
-instance
-  ( IR.TermAll (CanSubst ext primTy primVal) ext primTy primVal,
-    IR.ElimAll (CanSubst ext primTy primVal) ext primTy primVal
-  ) =>
-  CanSubst ext primTy primVal (IR.Elim' ext primTy primVal)
-  where
-  subst ii r (IR.Bound' j a)
-    | ii == j = r
-    | otherwise = IR.Bound' j (subst ii r a)
-  subst ii r (IR.Free' x a) = IR.Free' x (subst ii r a)
-  subst ii r (IR.Prim' k a) = IR.Prim' k (subst ii r a)
-  subst ii r (IR.App' it ct a) =
-    IR.App' (subst ii r it) (subst ii r ct) (subst ii r a)
-  subst ii r (IR.Ann' pi term t a) =
-    IR.Ann' pi (subst ii r term) (subst ii r t) (subst ii r a)
-  subst ii r (IR.ElimX a) = IR.ElimX (subst ii r a)
-
-substTerm ::
-  ( IR.TermAll (CanSubst ext primTy primVal) ext primTy primVal,
-    IR.ElimAll (CanSubst ext primTy primVal) ext primTy primVal
-  ) =>
-  Natural ->
-  IR.Elim' ext primTy primVal ->
-  IR.Term' ext primTy primVal ->
-  IR.Term' ext primTy primVal
+
+
+class HasWeak a where
```

Comment:
Should we put this in `Juvix.Library` perhaps?

Reply:
You mean the `HasWeak` class? Yeah, maybe, since it's not specific to core terms (unlike `subst` which has that `Elim` argument).